// @google-cloud/dataplex/build/src/index.js
"use strict";
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ** This file is automatically generated by synthtool. **
// ** https://github.com/googleapis/synthtool **
// ** All changes to this file may be overwritten. **
Object.defineProperty(exports, "__esModule", { value: true });
exports.protos = exports.MetadataServiceClient = exports.DataTaxonomyServiceClient = exports.DataScanServiceClient = exports.DataplexServiceClient = exports.ContentServiceClient = exports.v1 = void 0;
const v1 = require("./v1");
exports.v1 = v1;
const ContentServiceClient = v1.ContentServiceClient;
exports.ContentServiceClient = ContentServiceClient;
const DataplexServiceClient = v1.DataplexServiceClient;
exports.DataplexServiceClient = DataplexServiceClient;
const DataScanServiceClient = v1.DataScanServiceClient;
exports.DataScanServiceClient = DataScanServiceClient;
const DataTaxonomyServiceClient = v1.DataTaxonomyServiceClient;
exports.DataTaxonomyServiceClient = DataTaxonomyServiceClient;
const MetadataServiceClient = v1.MetadataServiceClient;
exports.MetadataServiceClient = MetadataServiceClient;
exports.default = {
    v1,
    ContentServiceClient,
    DataplexServiceClient,
    DataScanServiceClient,
    DataTaxonomyServiceClient,
    MetadataServiceClient,
};
const protos = require("../protos/protos");
exports.protos = protos;
