// @google-cloud/appengine-admin/build/src/v1/index.js
"use strict";
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ** This file is automatically generated by gapic-generator-typescript. **
// ** https://github.com/googleapis/gapic-generator-typescript **
// ** All changes to this file may be overwritten. **
Object.defineProperty(exports, "__esModule", { value: true });
exports.VersionsClient = exports.ServicesClient = exports.InstancesClient = exports.FirewallClient = exports.DomainMappingsClient = exports.AuthorizedDomainsClient = exports.AuthorizedCertificatesClient = exports.ApplicationsClient = void 0;
var applications_client_1 = require("./applications_client");
Object.defineProperty(exports, "ApplicationsClient", { enumerable: true, get: function () { return applications_client_1.ApplicationsClient; } });
var authorized_certificates_client_1 = require("./authorized_certificates_client");
Object.defineProperty(exports, "AuthorizedCertificatesClient", { enumerable: true, get: function () { return authorized_certificates_client_1.AuthorizedCertificatesClient; } });
var authorized_domains_client_1 = require("./authorized_domains_client");
Object.defineProperty(exports, "AuthorizedDomainsClient", { enumerable: true, get: function () { return authorized_domains_client_1.AuthorizedDomainsClient; } });
var domain_mappings_client_1 = require("./domain_mappings_client");
Object.defineProperty(exports, "DomainMappingsClient", { enumerable: true, get: function () { return domain_mappings_client_1.DomainMappingsClient; } });
var firewall_client_1 = require("./firewall_client");
Object.defineProperty(exports, "FirewallClient", { enumerable: true, get: function () { return firewall_client_1.FirewallClient; } });
var instances_client_1 = require("./instances_client");
Object.defineProperty(exports, "InstancesClient", { enumerable: true, get: function () { return instances_client_1.InstancesClient; } });
var services_client_1 = require("./services_client");
Object.defineProperty(exports, "ServicesClient", { enumerable: true, get: function () { return services_client_1.ServicesClient; } });
var versions_client_1 = require("./versions_client");
Object.defineProperty(exports, "VersionsClient", { enumerable: true, get: function () { return versions_client_1.VersionsClient; } });
