// @google-cloud/appengine-admin/build/src/index.js
"use strict";
// Copyright 2023 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     https://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// ** This file is automatically generated by synthtool. **
// ** https://github.com/googleapis/synthtool **
// ** All changes to this file may be overwritten. **
Object.defineProperty(exports, "__esModule", { value: true });
exports.protos = exports.VersionsClient = exports.ServicesClient = exports.InstancesClient = exports.FirewallClient = exports.DomainMappingsClient = exports.AuthorizedDomainsClient = exports.AuthorizedCertificatesClient = exports.ApplicationsClient = exports.v1 = void 0;
const v1 = require("./v1");
exports.v1 = v1;
const ApplicationsClient = v1.ApplicationsClient;
exports.ApplicationsClient = ApplicationsClient;
const AuthorizedCertificatesClient = v1.AuthorizedCertificatesClient;
exports.AuthorizedCertificatesClient = AuthorizedCertificatesClient;
const AuthorizedDomainsClient = v1.AuthorizedDomainsClient;
exports.AuthorizedDomainsClient = AuthorizedDomainsClient;
const DomainMappingsClient = v1.DomainMappingsClient;
exports.DomainMappingsClient = DomainMappingsClient;
const FirewallClient = v1.FirewallClient;
exports.FirewallClient = FirewallClient;
const InstancesClient = v1.InstancesClient;
exports.InstancesClient = InstancesClient;
const ServicesClient = v1.ServicesClient;
exports.ServicesClient = ServicesClient;
const VersionsClient = v1.VersionsClient;
exports.VersionsClient = VersionsClient;
exports.default = {
    v1,
    ApplicationsClient,
    AuthorizedCertificatesClient,
    AuthorizedDomainsClient,
    DomainMappingsClient,
    FirewallClient,
    InstancesClient,
    ServicesClient,
    VersionsClient,
};
const protos = require("../protos/protos");
exports.protos = protos;
